### DISCUSSION TITLE
make onSucces be a TS typeguard

### DISCUSSION THREAD

**Discussion by marekvanicek** [TanStack/query#5339] [Ideas]
So, `data` from useQuery is the `type | undefined`. I get that, but should not the data be always present if checked in onSuccess?

For example

```
function fetchData(): Promise<User[]> {...}
const {data, isSuccess} = useQuery{{queryFn: fetchUsers,..}}

if (isSuccess) {
    data // <- here the data is User[] | undefined. I feel like at this point it should be mandatory for the data to be present. 
    
}
```

My take on this is that isSuccess should be a **type guard**. Or introduce a new function that is a typeguard. Making the type in inside the isSucces if statement a User[].

**Comment by TkDodo** [comment]
> So, data from useQuery is the type | undefined

I mean ... that's really not what's happening. I can easily prove this ([typescript playground](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgVwM4FMCKz1QJ5wC+cAZlBCHAOQACMAhgHaoMDGA1gPRTr2swBaAI458VALAAoKTDxh0cAKoYocALyJgAEzgAuOI2QgARrkJSpJZI37AIjUuhisAFstyoAFAEp9ABXIQYAwAHncoAG0AXQA+RCk4OB4YZCgHAIpg9AA6HlQIABsAN3RPCKRtfQBGIijvKXNpSVZ7FkQtegYAGjhggGVkVlZ0VFRiDTQsUTxPJBFcPABpdDx9CKpJqFQqKJ75-AAxRn0SJ1dw1CJ6po6GKU5OOAA9AH4LSWASOE9+weHR7zxSSJRK3ej3R6vBIghpAA)), but it's tiring that you haven't done this yourself ...

**Reply by marekvanicek** [reply]
Yeah man, i suppose it is tiring. But Try switching the import into [vue-query](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgVwM4FMCKz1QJ5wC+cAZlBCHAOQACMAhgHaoMDGA1gPQBuOAtAEcc+KgFgAUBJh4w6OAFUMUOAF5EwACZwAXHEbIQAI1yEJEkskasYwCI1LoYrABaLcqABQBKXQAVyIMAYADxuUADaALoAfIgScHBQjshQ9v4UQegAdEmoEAA23Oge4UiaugCMRJFeEqaS4qx2LIga9AwANHBBAMrIrKzoqKjEamhYwngeSEK4eADS6Hi64VTjUKhUkV2z+ABijLokji5hqES1DW0MEpyccAB6APxm4sAkcB69-YPDWdz0fI4LxxcQJBLXei3e7POpAA). Then it works the way I described.

[EDIT] changed link -> forgot for .value on isSuccess initially

**Reply by TkDodo** [reply]
yep, that's why reproductions are important. You left out that you're using vue ;)

@DamianOsipiuk can you maybe have a look please? On query-core level, we have discriminated union types, but I think they might get lost in vue-land due to the unwrapping that happens?

**Reply by marekvanicek** [reply]
@TkDodo Yeah, fair enough. Vue-query is all I ever used therefore I thought this was expected behavior across all frameworks. My b!

**Reply by DamianOsipiuk** [reply]
It is a Vue limitation, but has a workaround.

https://github.com/TanStack/query/issues/5065#issuecomment-1458639556

**Reply by marekvanicek** [reply]
@DamianOsipiuk so what you are saying is we need to wrap a composable inside a reactive? That sure does sound hacky to me. Isnt there a better way to tackle this?

**Reply by DamianOsipiuk** [reply]
There is.  
We need to make a breaking change and return `reactive` object instead of object with `refs`.
This in turn will brak reactivity whenever any dev will destructure result like `const {data, isSuccess} = useQuery{{queryFn: fetchUsers,..}}` which is exactly what you have tried.  
In this case you will stop getting updates. And i think this will be a common error among devs moving from other frameworks like `react`.
I prefer to keep reactivity more resilient to dev error, than to give you seamless typescript experience.  
As typescript can be easily fixed while reactivity problems could be hard to track.
